File: 2.OOP/concepts/MediumExample/design_patterns/Structural/Flyweight/Flyweight.go
package main

import (
	"fmt"
)

type Font interface {
	GetName() string
}
type DefaultFont struct{}

func (f *DefaultFont) GetName() string {
	return "Default"
}

type BoldFont struct{}

func (f *BoldFont) GetName() string {
	return "Bold"
}

type FontFactory struct {
	fonts map[string]Font
}

func NewFontFactory() *FontFactory {
	return &FontFactory{fonts: make(map[string]Font)}
}

func (f *FontFactory) GetFont(fontName string) Font {
	if font, ok := f.fonts[fontName]; ok {
		return font
	}
	switch fontName {
	case "Default":
		f.fonts[fontName] = &DefaultFont{}
	case "Bold":
		f.fonts[fontName] = &BoldFont{}
	default:
		panic("Font not supported: " + fontName)
	}
	return f.fonts[fontName]
}

type Text struct {
	content string
	font    Font
}

func NewText(content string, factory *FontFactory) *Text {
	return &Text{content: content, font: factory.GetFont("Default")} // Default font
}

func (t *Text) Render() string {
	return t.font.GetName() + ": " + t.content
}

func main() {
	factory := NewFontFactory()

	text1 := NewText("This is some text", factory)
	text2 := NewText("This is bold text", factory)
	text2.font = factory.GetFont("Bold") // Reuse existing font

	fmt.Println(text1.Render()) // Output: Default: This is some text
	fmt.Println(text2.Render()) // Output: Bold: This is bold text
}

File: 2.OOP/concepts/MediumExample/design_patterns/Structural/Flyweight/Flyweight.py
from abc import ABC,abstractmethod
# Flyweight Interface (Abstract Class):
class Font(ABC):
    @abstractmethod
    def get_name(self)-> str:
        pass

# concrete flyweights (Fonts)
class DefaultFont(Font):
    def get_name(self)->str:
        return "Default"

class BoldFont(Font):
    def get_name(self)-> str:
        return "Bold"

class FontFactory:
    def __init__(self):
        self.fonts = {}
    def get_font(self, font_name: str) -> Font:
        if font_name not in self.fonts:
            if font_name == "Default":
                self.fonts[font_name] = DefaultFont()
            elif font_name == "Bold":
                self.fonts[font_name] = BoldFont()
            else:
                raise ValueError("Font not supported:", font_name)
        return self.fonts[font_name]


class Text :
    def __init__(self,content:str,font_factory:FontFactory):
        self.content = content
        self.font = font_factory.get_font("Default")  # Default font

    def render(self) -> str:
        return f"{self.font.get_name()}: {self.content}"

factory = FontFactory()

text1 = Text("This is some text", factory)
text2 = Text("This is bold text", factory)
text2.font = factory.get_font("Bold")

print(text1.render())
print(text2.render())